pyCalculations/fieldtracer.py

```diff
@@ -190,89 +190,140 @@ def static_field_tracer_3d( vlsvReader, coord_list, max_iterations, dx, directio
+      z = np.arange(mins[2], maxs[2], dcell[2]) + 0.5*dcell[2]
+      coordinates = np.array([x,y,z], dtype=object)
+
+      # Create grid interpolation of vector field (V)
+      interpolator_face_V_0 = interpolate.RegularGridInterpolator((x-0.5*dcell[0], y, z), fg[:,:,:,0], bounds_error = False, fill_value = np.nan)
+      interpolator_face_V_1 = interpolate.RegularGridInterpolator((x, y-0.5*dcell[1], z), fg[:,:,:,1], bounds_error = False, fill_value = np.nan)
+      interpolator_face_V_2 = interpolate.RegularGridInterpolator((x, y, z-0.5*dcell[2]), fg[:,:,:,2], bounds_error = False, fill_value = np.nan)
+      interpolators = [interpolator_face_V_0, interpolator_face_V_1, interpolator_face_V_2]
+
+      # Trace vector field lines
+      points_traced = [np.array(coord_list)]              # iteratively append traced trajectories to this list
+      points = points_traced[0]
+      N = len(list(coord_list))
+      V_unit = np.zeros([3, N])
+      for i in range(max_iterations):
+         V_unit[0, :] = interpolators[0](points)
+         V_unit[1, :] = interpolators[1](points)
+         V_unit[2, :] = interpolators[2](points)
+         V_mag = np.linalg.norm(V_unit, axis=(0))
+         V_unit = V_unit / V_mag[np.newaxis,:]
+         new_points = points + multiplier*V_unit.T * dx
+         points = new_points
+         points_traced.append( list(points) )             # list of lists of 3-element arrays
+   
+   elif vg is not None:
+      # Search for the unique coordinates in the given seeds only
+      unique_seed_coords,indices = np.unique(seed_coords, axis = 0, return_inverse = True)    # indice here is to reverse the coords order to initial
+      n_unique_seeds = unique_seed_coords.shape[0]
+      points_traced_unique = np.zeros((n_unique_seeds, max_iterations, 3))
+      print(vg)
+      def find_unit_vector(vg, coord):
+         val_at_point = vlsvReader.read_interpolated_variable(vg,coord)
+         val_mag = np.linalg.norm(val_at_point, axis = 1, keepdims = True)
+         return val_at_point/val_mag
+      
+      unique_seed_coords,indices = np.unique(seed_coords, axis = 0, return_inverse = True)    # indice here is to reverse the coords order to initial
+      n_unique_seeds = unique_seed_coords.shape[0]
+      points_traced_unique = np.zeros((n_unique_seeds, max_iterations, 3))
+
+      Re = 6371000
+      mask_update = np.ones((n_unique_seeds,),dtype = bool) # A mask to determine if the points are still needed to trace further
+      points_traced_unique[:, 0, :] = unique_seed_coords
+
+      for i in range(1, max_iterations):
+
+         var_unit = find_unit_vector(vg, points_traced_unique[:, i-1, :])
+         next_points = points_traced_unique[:, i-1, :] + multiplier * dx * var_unit
+
+         points_traced_unique[mask_update,i,:] = next_points[mask_update,:]
+
+         distances = np.linalg.norm(points_traced_unique[:,i,:],axis = 1)
+         mask_update[(distances <= 5*Re) | (distances >= 40*Re)] = False
```

Comment:
`(distances <= 5*Re) | (distances >= 40*Re)` should rather be a function that can be passed to the tracer.

So that the function passed accepts the array of points and tells the tracer that these traces can be stopped.